Automattic/newspack-ads
# feat(marketplace): purchase sponsorship

Implements basic ad order definitions and integration with GAM.

| WooCommerce Order | GAM Order Line Item |
| --- | --- |
| <img width="847" alt="image" src="https://user-images.githubusercontent.com/820752/236931177-f05736c2-b1aa-45cf-bb8a-b4ae0e63765d.png"> | <img width="799" alt="image" src="https://user-images.githubusercontent.com/820752/236931094-aa6ee92a-bd7b-415b-95a6-bcd3f7f5d600.png"> |

An ad product order purchase should include:
 - From date
 - To date
 - Total days

We'll be using a temporary block called "Marketplace" for offering and purchasing ad products, which after product selection will redirect to checkout.

## How to test

1. Check out this branch and [newspack-plugin/epic/ads-marketplace](https://github.com/Automattic/newspack-plugin/tree/epic/ads-marketplace)
2. Make sure you are connected to a GAM account
3. If you don't have any Ad Products created, visit the Advertising -> Marketplace page and create some ad products (more details https://github.com/Automattic/newspack-ads/pull/566)
4. Edit a new page and add the "Marketplace" block, which renders only a placeholder component in the editor
5. Visit the page and confirm you see the products with date inputs:

<img width="480" alt="image" src="https://user-images.githubusercontent.com/820752/236933677-3dd5d33e-b44e-44cf-955f-e268e12f7bdb.png">

6. Select a product, a valid date range (cannot be starting Today), and proceed to payment
7. Finish checkout and confirm the order is created in your local dashboard
8. Visit GAM, locate the created order and confirm the created line item data matches the ad product and order details:
   1. The creative placeholders matches the product
   2. The date range matches 12am of the "from" date and 11:59pm of the "to" date
   3. Total pricing should also be the same
9. The WooCommerce order should also have the values for the meta `newspack_ads_gam_network_code` and `newspack_ads_gam_order_id`, which will later be used for creating links for quick access to the GAM order

Added tests for validating cart item data should also pass.
